File: lib/nb_pingcrm_web/serializers/contact_serializer.ex
defmodule NbPingcrmWeb.Serializers.ContactSerializer do
  @moduledoc false
  use NbSerializer.Serializer

  alias NbPingcrm.CRM.{Contact, Organization}

  schema do
    field(:id, :number)
    field(:first_name, :string)
    field(:last_name, :string)
    field(:name, :string, compute: :compute_name)
    field(:email, :string, nullable: true)
    field(:phone, :string, nullable: true)
    field(:address, :string, nullable: true)
    field(:city, :string, nullable: true)
    field(:region, :string, nullable: true)
    field(:country, :string, nullable: true)
    field(:postal_code, :string, nullable: true)
    field(:deleted_at, :datetime, nullable: true)
    field(:organization_id, :number, nullable: true)
    field(:organization_name, :string, compute: :compute_organization_name, nullable: true)
  end

  def compute_name(contact, _opts) do
    Contact.name(contact)
  end

  def compute_organization_name(%{organization: %Organization{name: name}}, _opts) do
    name
  end

  def compute_organization_name(_, _opts), do: nil
end

File: assets/js/types/ContactSerializer.ts
export interface Contact {
  address: string | null;
  city: string | null;
  country: string | null;
  deletedAt: string | null;
  email: string | null;
  firstName: string;
  id: number;
  lastName: string;
  name: string;
  organizationId: number | null;
  organizationName: string | null;
  phone: string | null;
  postalCode: string | null;
  region: string | null;
}

import { z } from 'zod';

export const ContactSchema = z.object({
  address: z.string().nullable(),
  city: z.string().nullable(),
  country: z.string().nullable(),
  deletedAt: z
    .codec(z.iso.datetime({ offset: true }), z.date(), {
      decode: (value) => {
        const date = new Date(value);
        if (isNaN(date.getTime())) throw new Error('Invalid ISO datetime');
        return date;
      },
      encode: (value) => value.toISOString(),
    })
    .nullable(),
  email: z.string().nullable(),
  firstName: z.string(),
  id: z.number(),
  lastName: z.string(),
  name: z.string(),
  organizationId: z.number().nullable(),
  organizationName: z.string().nullable(),
  phone: z.string().nullable(),
  postalCode: z.string().nullable(),
  region: z.string().nullable(),
});

/** Wire/input representation accepted by ContactSchema. */
export type ContactWire = z.input<typeof ContactSchema>;

/** Runtime/output representation returned by ContactSchema. */
export type ContactRuntime = z.output<typeof ContactSchema>;
